core/rest-client/src/http_client.rs

```diff
@@ -0,0 +1,425 @@
+	use std::vec::Vec;
+
+	#[test]
+	fn join_url_adds_query_parameters() {
+		let base_url = Url::parse("https://example.com").unwrap();
+		let path = "api/v2/example_list";
+		let query = [("filter", "all"), ("order", ("desc"))];
+
+		let complete_url = join_url(base_url, path, Some(&query)).unwrap();
+
+		assert_eq!(
+			complete_url.as_str(),
+			"https://example.com/api/v2/example_list?filter=all&order=desc"
+		);
+	}
+
+	#[test]
+	fn join_url_has_no_query_parameters() {
+		let base_url = Url::parse("https://example.com").unwrap();
+		let path = "api/v2/endpoint";
+		let complete_url = join_url(base_url, path, None).unwrap();
+		assert_eq!(complete_url.as_str(), "https://example.com/api/v2/endpoint");
+	}
+
+	#[test]
+	fn join_url_with_too_many_slashes() {
+		let base_url = Url::parse("https://api.mydomain.com").unwrap();
+		let path = "/api/v1/post";
+		let complete_url = join_url(base_url, path, None).unwrap();
+		assert_eq!(complete_url.as_str(), "https://api.mydomain.com/api/v1/post");
+	}
+
+	// #[test]
+	// fn get_from_mock_api() {
+	// 	// Data structure that matches with REST API JSON
+	// 	#[derive(Serialize, Deserialize, Debug)]
+	// 	struct MockMessage {
+	// 		pub message: String,
+	// 	}
+	//
+	// 	impl RestPath<()> for MockMessage {
+	// 		fn get_path(_: ()) -> Result<String, Error> {
+	// 			Ok(format!("message"))
+	// 		}
+	// 	}
+	//
+	// 	let http_client = HttpClient::new(true, Some(Duration::from_secs(3u64)), None, None);
+	// 	let base_url = Url::parse("https://integritee.free.beeceptor.com").unwrap();
+	//
+	// 	let (response, encoded_body) = http_client
+	// 		.send_request::<(), MockMessage>(base_url, Method::GET, (), None, None)
+	// 		.unwrap();
+	//
+	// 	let response_body: MockMessage =
+	// 		deserialize_response_body(encoded_body.as_slice()).unwrap();
+	//
+	// 	assert!(response.status_code().is_success());
+	// 	assert!(!response_body.message.is_empty());
+	// }
+
+	#[test]
+	fn get_with_parameters() {
+		#[derive(Serialize, Deserialize, Debug)]
+		struct RequestArgs {
+			pub order: String,
+			pub filter: String,
+		}
+
+		// Data structure that matches with REST API JSON
+		#[derive(Serialize, Deserialize, Debug)]
+		struct HttpBinAnything {
+			pub args: RequestArgs,
+			pub origin: String,
+			pub url: String,
+		}
+
+		impl RestPath<()> for HttpBinAnything {
+			fn get_path(_: ()) -> Result<String, Error> {
+				Ok(format!("anything"))
+			}
+		}
+
+		let http_client = HttpClient::new(
+			true,
+			Some(Duration::from_secs(3u64)),
+			Some(headers_connection_close()),
+			None,
+		);
+		let base_url = Url::parse("https://httpbin.org").unwrap();
+		let query_parameters = [("order", "desc"), ("filter", "all")];
+
+		let (response, encoded_body) = http_client
+			.send_request::<(), HttpBinAnything>(
+				base_url,
+				Method::GET,
+				(),
+				Some(&query_parameters),
+				None,
+			)
+			.unwrap();
+
+		let response_body: HttpBinAnything =
+			deserialize_response_body(encoded_body.as_slice()).unwrap();
```

Comment:
We send a request to `httpbin.org`, which is a HTTP/REST test API. That makes it not really a unit test anymore, since we have to rely on a connection to the outside and that service to be running. 
In the future we could host this test service ourselves, in a docker, but I'm not sure it's worth the effort.

Reply:
You are right about that. Since it only needs internet connection, I think it's fine for now.